### PR TITLE
Deprecate use_batchnorm in favor of generalized use_norm parameter

### PR DESCRIPTION
**Pull request by GuillaumeErhard**
Hello,

a bit late to implement my suggestion in #983 but here it is. Open to any suggestion to make it clearer / simpler or any wanted test.

### CODE REVIEW

File: segmentation_models_pytorch/base/modules.py [hunk 1]
```diff
@@ -16,11 +18,53 @@ def __init__(
...
+        elif not isinstance(use_norm, dict):
+            raise ValueError("use_norm must be a dictionary, boolean, or string")
+
+        if use_norm["type"] == "inplace" and InPlaceABN is None:
             raise RuntimeError(
-                "In order to use `use_batchnorm='inplace'` inplace_abn package must be installed. "
-                + "To install see: https://github.com/mapillary/inplace_abn"
+                "In order to use `use_batchnorm='inplace'` or `use_norm='inplace'` the inplace_abn package must be installed. "
+                "To install see: https://github.com/mapillary/inplace_abn"
             )
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:68]: Let's have a separate function `get_norm_layer` which will validate input params and return norm layer

File: segmentation_models_pytorch/base/modules.py [hunk 2]
```diff
@@ -127,3 +180,9 @@ def __init__(self, name, **params):
 
     def forward(self, x):
         return self.attention(x)
+
+
+if __name__ == "__main__":
+    print(Conv2dReLU(3, 12, 4))
+    print(Conv2dReLU(3, 12, 4, use_norm={"type": "batchnorm"}))
+    print(Conv2dReLU(3, 12, 4, use_norm={"type": "layernorm", "eps": 1e-3}))
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:188]: Should be removed, instead, it would be nice to add a test

File: segmentation_models_pytorch/decoders/linknet/decoder.py [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
 
 
 class TransposeX2(nn.Sequential):
-    def __init__(self, in_channels: int, out_channels: int, use_batchnorm: bool = True):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        use_batchnorm: Union[bool, str, None] = True,
+        use_norm: Union[bool, str, Dict[str, Any]] = True,
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/linknet/decoder.py:14]: We should have only `use_norm` here, use_bathcnorm should be replaced on top level (e.g. Unet model)

File: segmentation_models_pytorch/decoders/linknet/decoder.py [hunk 4]
```diff
@@ -15,14 +21,20 @@ def __init__(self, in_channels: int, out_channels: int, use_batchnorm: bool = Tr
...
 
 class DecoderBlock(nn.Module):
-    def __init__(self, in_channels: int, out_channels: int, use_batchnorm: bool = True):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        use_batchnorm: Union[bool, str, None] = True,
+        use_norm: Union[bool, str, Dict[str, Any]] = True,
+    ):
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/linknet/decoder.py:37]: same here, let's move it to the model, and leave only `use_norm`

**Review by qubvel — COMMENTED** [review]
Hi @GuillaumeErhard, huge thanks for taking time and working on the PR! Here a few comments

File: segmentation_models_pytorch/base/modules.py (hunk 1 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: Good catch. Much simpler. Done the changes

File: segmentation_models_pytorch/base/modules.py (hunk 2 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: Yeap, added some test around Conv2dReLu

File: segmentation_models_pytorch/decoders/linknet/decoder.py (hunk 3 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/linknet/decoder.py]: Made a proposition

File: segmentation_models_pytorch/decoders/linknet/decoder.py (hunk 4 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/linknet/decoder.py]: Made a proposition

**Comment by GuillaumeErhard** [comment]
I am missing some test of before and after on calling model using batchnorm and using norm to show that the model are equivalent. I will do this tomorrow. ( tests/encoders/test_batchnorm_deprecation.py )

But I have hit an error that I did not anticipate with the test. (Thank god for test driven dev). GroupNorm uses num_groups, num_channels param instead of using a simple out_channels As of right now I don't see a good way to make it work so I am thinking of dropping the support for it. If you have any idea, I am open for it

**Comment by GuillaumeErhard** [comment]
Added the test so that we get the same model before and after the changes. Which showed that Linknet additional changes which are now done.

I respected that all element in model/decoder can still work use_norm can be str / bool / dict for the moment in the case people use them directly. For that I need to sanitize at the model level and on Conv2DRelu. Tell me if you want me to change to dict only in decoder

I also saw that pspnet had also batchnorm usage but through psp_use_batchnorm and aligned it with use_norm.
By the way maybe on line 20 of pspnet/decoder.py it can be changed for a norm instead of identity ?

Removed groupnorm for the moment. Taking any suggestion to make it possible. See problem described above

Added support in upernet to also use_norm.
Is the behavior on l 35 expected ? That is always batchnorm ? Or I can pass use_norm also ?

File: segmentation_models_pytorch/base/modules.py [hunk 5]
```diff
@@ -6,6 +9,74 @@
...
+    elif norm_type == "identity":
+        norm = nn.Identity()
+    elif norm_type == "layernorm":
+        norm = nn.LayerNorm(out_channels, **extra_kwargs)
+    elif norm_type == "instancenorm":
+        norm = nn.InstanceNorm2d(out_channels, **extra_kwargs)
+    else:
+        raise ValueError(f"Unrecognized normalization type: {norm_type}")
+
+    return norm
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:79]: I would rather put everything into a single simple function as follows:
```
if use_norm is True:
    params = {"type": "batchnorm"}
elif use_norm is False:
    params = {"type": "identity"}
elif use_norm == "inplace":
    params = {"type": "inplace", "activation": "leaky_relu", "activation_param": 0.0}

if not isinstance(params, dict) -> raise
if not "type" in params -> raise
if not params["type"] in supproted_norms -> raise

<dispatch to norm layer>

return norm 
```

File: segmentation_models_pytorch/base/modules.py [hunk 6]
```diff
@@ -15,12 +86,13 @@ def __init__(
...
+        use_norm="batchnorm",
     ):
-        if use_batchnorm == "inplace" and InPlaceABN is None:
+        use_norm = normalize_use_norm(use_norm)
+        if use_norm["type"] == "inplace" and InPlaceABN is None:
             raise RuntimeError(
-                "In order to use `use_batchnorm='inplace'` inplace_abn package must be installed. "
-                + "To install see: https://github.com/mapillary/inplace_abn"
+                "In order to use `use_batchnorm='inplace'` or `use_norm='inplace'` the inplace_abn package must be installed. "
+                "To install see: https://github.com/mapillary/inplace_abn"
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:95]: Also not needed here, should be under get_norm_layer function

File: segmentation_models_pytorch/base/modules.py [hunk 7]
```diff
@@ -29,21 +101,16 @@ def __init__(
             kernel_size,
             stride=stride,
             padding=padding,
-            bias=not (use_batchnorm),
+            bias=use_norm["type"] != "inplace",
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:104]: We can initialize norm first and use a separate varaible

`is_inplace_batchnorm = norm.__name__ == "InPlaceABN"`

File: segmentation_models_pytorch/base/modules.py [hunk 8]
```diff
@@ -29,21 +101,16 @@ def __init__(
...
         )
-        relu = nn.ReLU(inplace=True)
+        norm = get_norm_layer(use_norm, out_channels)
 
-        if use_batchnorm == "inplace":
-            bn = InPlaceABN(out_channels, activation="leaky_relu", activation_param=0.0)
+        if use_norm["type"] == "inplace":
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:108]: same here

File: segmentation_models_pytorch/decoders/linknet/model.py [hunk 9]
```diff
@@ -60,7 +79,8 @@ def __init__(
         encoder_name: str = "resnet34",
         encoder_depth: int = 5,
         encoder_weights: Optional[str] = "imagenet",
-        decoder_use_batchnorm: bool = True,
+        decoder_use_batchnorm: Union[bool, str, None] = None,
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/linknet/model.py:82]: let's remove`decoder_use_batchnorm` from signature and pop from krawgs later

File: segmentation_models_pytorch/decoders/linknet/model.py [hunk 10]
```diff
@@ -82,11 +102,12 @@ def __init__(
             **kwargs,
         )
 
+        decoder_use_norm = normalize_decoder_norm(decoder_use_batchnorm, decoder_use_norm)
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/linknet/model.py:105]: here we just resolve the name

File: segmentation_models_pytorch/decoders/manet/model.py [hunk 11]
```diff
@@ -63,7 +82,8 @@ def __init__(
         encoder_name: str = "resnet34",
         encoder_depth: int = 5,
         encoder_weights: Optional[str] = "imagenet",
-        decoder_use_batchnorm: bool = True,
+        decoder_use_batchnorm: Union[bool, str, None] = None,
+        decoder_use_norm: Union[bool, str, Dict[str, Any]] = "batchnorm",
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/manet/model.py:86]: Let's keep default

`decoder_use_norm: Union[bool, str, Dict[str, Any]] = True`

and remove `decoder_use_batchnorm` entirely from signature, it will be passed in kwargs and poped from there

File: tests/base/test_modules.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
+from segmentation_models_pytorch.base.modules import Conv2dReLU
+
+
+def test_conv2drelu_batchnorm():
+    module = Conv2dReLU(3, 16, kernel_size=3, padding=1, use_norm="batchnorm")
+
+    expected = ('Conv2dReLU(\n  (0): Conv2d(3, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))'
+                '\n  (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)'
+                '\n  (2): ReLU(inplace=True)\n)')
+    assert repr(module) == expected
```
Comment (qubvel) [inline tests/base/test_modules.py:10]: Not super robust, formatting can changed in any version, lets use

`assert isinstance(module[1], nn.BatchNorm2d)`

and below please

File: tests/encoders/test_batchnorm_deprecation.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        model_decoder_batchnorm = create_model(
+            "pspnet",
+            "mobilenet_v2",
+            None,
+            psp_use_batchnorm=decoder_option
+        )
+    torch.manual_seed(42)
+    model_decoder_norm = create_model("pspnet", "mobilenet_v2", None, psp_use_batchnorm=None, decoder_use_norm=decoder_option)
+
+    check_two_models_strictly_equal(model_decoder_batchnorm, model_decoder_norm)
```
Comment (qubvel) [inline tests/encoders/test_batchnorm_deprecation.py:40]: Nice!

File: tests/utils.py [hunk 14]
```diff
@@ -58,3 +58,10 @@ def check_run_test_on_diff_or_main(filepath_patterns: List[str]):
                 return True
 
     return False
+
+
+def check_two_models_strictly_equal(model_a: torch.nn.Module, model_b: torch.nn.Module) -> None:
+    for (k1, v1), (k2, v2) in zip(model_a.state_dict().items(),
+                                  model_b.state_dict().items()):
+        assert k1 == k2, f"Key mismatch: {k1} != {k2}"
+        assert (v1 == v2).all(), f"Tensor mismatch at key '{k1}':\n{v1} !=\n{v2}"
```
Comment (qubvel) [inline tests/utils.py:67]: I would also add forward pass here and compare logits

**Review by qubvel — COMMENTED** [review]
Thanks for addressing comments, please see the review 🤗

File: segmentation_models_pytorch/base/modules.py (hunk 5 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: So no str for default usage like LayerNorm etc... ??? I kept it for the moment. Tell me if I have to remove.

Did changed accordingly and tried to match the proposed flow as closely as possible

File: segmentation_models_pytorch/base/modules.py (hunk 6 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: Removed and placed in get_norm_layer

File: segmentation_models_pytorch/base/modules.py (hunk 7 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: yeap simpler

File: segmentation_models_pytorch/base/modules.py (hunk 8 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: done

File: segmentation_models_pytorch/decoders/linknet/model.py (hunk 9 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/linknet/model.py]: removed decoder_use_batchnorm and used kwargs. I like this approach

File: segmentation_models_pytorch/decoders/linknet/model.py (hunk 10 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/linknet/model.py]: Tell me if this is what you had in mind

File: segmentation_models_pytorch/decoders/manet/model.py (hunk 11 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/manet/model.py]: I am not really that sold with True as default. "batchnorm" seems more meaningful and the bool more of an historic usage. 
Changed it to True on model. Tell me if I have to propagate in all the decoder

removed decoder_use_batchnorm and used kwargs. I like this approach

File: tests/base/test_modules.py (hunk 12 above)
Comment (GuillaumeErhard) [reply tests/base/test_modules.py]: Changed asserts

File: tests/utils.py (hunk 14 above)
Comment (GuillaumeErhard) [reply tests/utils.py]: Add input_data param to check that

File: segmentation_models_pytorch/decoders/linknet/model.py (hunk 10 above)
Comment (qubvel) [reply segmentation_models_pytorch/decoders/linknet/model.py]: > "batchnorm" seems more meaningful and the bool more of an historic usage.

Ok, sounds good!

File: segmentation_models_pytorch/base/modules.py [hunk 15]
```diff
@@ -6,6 +8,64 @@
...
+                "layernorm",
+                "instancenorm",
+        ):
+            norm_params = {"type": norm_str}
+        else:
+            raise ValueError(f"Unrecognized normalization type string provided: {use_norm}. Should be in {supported_norms}")
+    elif isinstance(use_norm, dict):
+        norm_params = use_norm
+    else:
+        raise ValueError("use_norm must be a dictionary, boolean, or string. Please refer to the documentation.")
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:39]: Let's have a more descriptive error here, I mean specify what kind of string and dict structure it should be.

File: segmentation_models_pytorch/base/modules.py [hunk 16]
```diff
@@ -6,6 +8,64 @@
...
+            }
+        elif norm_str in (
+                "batchnorm",
+                "identity",
+                "layernorm",
+                "instancenorm",
+        ):
+            norm_params = {"type": norm_str}
+        else:
+            raise ValueError(f"Unrecognized normalization type string provided: {use_norm}. Should be in {supported_norms}")
```
Comment (qubvel) [inline segmentation_models_pytorch/base/modules.py:35]: Nice!

File: segmentation_models_pytorch/decoders/linknet/model.py [hunk 17]
```diff
@@ -29,9 +30,22 @@ class Linknet(SegmentationModel):
...
+            - **dict**: Fully customizable normalization settings. Structure:
+              ```python
+              {"type": <norm_type>, **kwargs}
+              ```
+              where `norm_name` corresponds to normalization type (see above), and `kwargs` are passed directly to the normalization layer as defined in PyTorch documentation.
+
+            **Example**:
+            ```python
+            use_norm={"type": "layernorm", "eps": 1e-2}
+            ```
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/linknet/model.py:48]: Thanks for the detailed docstring, really appretiate it!

File: segmentation_models_pytorch/decoders/pspnet/model.py [hunk 18]
```diff
@@ -28,9 +29,22 @@ class PSPNet(SegmentationModel):
...
+              `"batchnorm"`, `"identity"`, `"layernorm"`, `"instancenorm"`, `"inplace"`.
+            - **dict**: Fully customizable normalization settings. Structure:
+              ```python
+              {"type": <norm_type>, **kwargs}
+              ```
+              where `norm_name` corresponds to normalization type (see above), and `kwargs` are passed directly to the normalization layer as defined in PyTorch documentation.
+
+            **Example**:
+            ```python
+            use_norm={"type": "layernorm", "eps": 1e-2}
```
Comment (qubvel) [inline segmentation_models_pytorch/decoders/pspnet/model.py:46]:
```suggestion
            decoder_use_norm={"type": "layernorm", "eps": 1e-2}
```

**Review by qubvel — APPROVED** [review]
Thanks, just a few minor comments and we are good to merge!

File: segmentation_models_pytorch/decoders/manet/model.py (hunk 11 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/manet/model.py]: Changed to "batchnorm"

File: segmentation_models_pytorch/decoders/pspnet/model.py (hunk 18 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/decoders/pspnet/model.py:46]: good catch

File: segmentation_models_pytorch/base/modules.py (hunk 15 above)
Comment (GuillaumeErhard) [reply segmentation_models_pytorch/base/modules.py]: Made a proposition

**Comment by qubvel** [comment]
Thanks for your contribution! I pushed some minor fixed to merge this 🤗
